cs/src/Connections/TunnelClient.cs

```diff
@@ -254,6 +258,43 @@ private void ConfigurePortForwardingService()
         {
             pfs.MessageFactory = this;
             pfs.ForwardedPortConnecting += OnForwardedPortConnecting;
+            pfs.RemoteForwardedPorts.PortAdded += (_, e) => OnForwardedPortAdded(pfs, e);
+            pfs.RemoteForwardedPorts.PortUpdated += (_, e) => OnForwardedPortAdded(pfs, e);
+        }
+    }
+
+    private void OnForwardedPortAdded(PortForwardingService pfs, ForwardedPortEventArgs e)
+    {
+        var port = e.Port.RemotePort;
+        if (!port.HasValue)
+        {
+            return;
+        }
+
+        var disconnectedStreamsCount = 0;
+        lock (this.disconnectedStreams)
+        {
+            // If there are disconnected streams for the port, re-connect them now.
+            if (this.disconnectedStreams.TryGetValue(port.Value, out var streams))
+            {
+                disconnectedStreamsCount = streams.Count;
+            }
+        }
+
+        for (int i = 0; i < disconnectedStreamsCount; i++)
+        {
+            Task.Run(async () =>
+            {
+                try
+                {
+                    await pfs.ConnectToForwardedPortAsync(port.Value, CancellationToken.None);
+                }
+                catch (Exception ex)
+                {
+                    this.Trace.Warning(
+                        $"Failed to reconnect to forwarded port {port}: {ex.Message}");
```

Comment:
What would happen if 1 of 5 streams failed to reconnect? I'm wondering if we should just keep trying `pfs.ConnectToForwardedPortAsync()` until we either succeed or encounter X failures. (Perhaps it would be such a rare event for a partial failure that it's not worth it?)

Reply:
If one of the streams fails to reconnect, then after a 60s timeout the application using the stream would eventually see it as a socket disconnection.

I don't think retrying here is likely to be helpful. If the connection to the forwarded port fails that means the host application is no longer listening on that port on the host. So, actually in that case we should immediately close any _disconnected_ streams for that port, so client applications don't have to wait for a timeout. I'll make that code change.

Reply:
Updated.